File: 프로그래머스/1/468371. 노란불 신호등/Hyun-suck Lee/노란불 신호등.java
// 1차 수정 - 태성씨 코드 보고 time 기준 for문이 더 적합하다고 판단해서 재작성
// 2차 수정 - LCM 공부 후 MAX값 설정
import java.util.*;

class Solution {
    public int solution(int[][] signals) {
        // LCM 구해서 max에 할당
        List<Integer> signalSums = new ArrayList<>();
        for (int[] signal : signals) {
            signalSums.add(this.getArraySum(signal));
        }
        int max = this.getLCM(signalSums),
            signalCount = signals.length;

        // 시간 기준으로 순차 탐색
        for (int time = 1; time <= max; time++) {
            int yellowCount = 0;
  
            for (int[] signal : signals) {
                int green  = signal[0],
                    yellow = signal[1],
                    red    = signal[2],
                    all    = green + yellow + red;
                
                int target = time % all;
                // yellow 범위면
                if (green < target && target <= green + yellow) {
                    yellowCount++;
                } else {
                    break;
                }
            }
            
            if (yellowCount == signalCount) {
                return time;
            }
        }
  
        return -1;
    }
    
    // GCD = (큰 수 % 작은수) -> (작은수 % value_1) -> (value_1 % value_2) ... 가 0이 되는 나눈 수 
    private int getGCD(int num1, int num2) {
        while (num2 != 0) {
            int temp = num1 % num2;
            num1 = num2;
            num2 = temp;
        }
        
        return num1;
    }
    
    // LCM = (A * B) / GCD_최대공약수
    private int getLCM(List<Integer> list) {
        int lcm = list.get(0);
        
        for (int idx = 1; idx < list.size(); idx++) {
            int num = list.get(idx);
            lcm = (lcm * num) / this.getGCD(lcm, num);
        }
        
        return lcm;
    }
    
    private int getArraySum(int[] array) {
        int res = 0;
        for (int num : array) {
            res += num;
        }
        return res;
    }
} 

/* 학습 (코드 기준)
LCM = (A * B) / GCD_최대공약수
GCD = (큰 수 % 작은수) -> (작은수 % value_1) -> (value_1 % value_2) ... 가 0이 되는 나눈 수 


*/

/*
n: 도로에 배치된 차량 신호등 개수
신호등 순서: green -> yellow -> red  
지속 시간: 신호등 마다 다름 (1sec ~ 18sec)
처음: 무조건 green (signals)

G+Y+R 은 20보다 작음 400까지만 하면 되나?

return: 모든 신호등이 노란불이 되는 가장 빠른 시각 or -1


2/1/1 이면
1 2 [3] 4 5 6 [7] 8 9 10 [11] ...

7 % 4 = 3
green(2) + yellow(1) = 3

3/2/4 이면
1 2 3 [4] [5] 6 7 8 9 10 11 12 [13] [14] 15 16 17 18...

1 + green ~ yellow 반복
+ red

*/

File: 프로그래머스/1/468371. 노란불 신호등/HONG/노란불 신호등.cpp
#include <string>
#include <vector>

using namespace std;

int gcd(int a, int b) {
    while(b != 0) {
        int temp = a % b;
        a = b;
        b = temp;
    }
    return a;
}

int lcm(int a, int b) {
    return a * b / gcd(a, b);
}

int solution(vector<vector<int>> signals) {
    int answer = 0, max = 1;
    int x, i;
    int total_signals_sec, start_yellow_sec, yellow_signal_sec;
    for(i=0;i<signals.size();i++) {
        total_signals_sec = signals[i][0] + signals[i][1] + signals[i][2];
        max = lcm(max, total_signals_sec);
    }
    
    for(x=0;x<max;x++) {
        for(i=0;i<signals.size();i++) {
            total_signals_sec = signals[i][0] + signals[i][1] + signals[i][2];
            start_yellow_sec = signals[i][0] + 1;
            yellow_signal_sec = x % total_signals_sec;
            
            if(yellow_signal_sec < start_yellow_sec || yellow_signal_sec > start_yellow_sec + signals[i][1] - 1)
                break;
         
        }
        if(i==signals.size()) {
            answer = x;
            break;
        } else {
            answer = -1;
        }
    }
    return answer;
}

/*
- signal 1
처음 시작 값 : 3초
노란불 : +5초
13 % 5 = 3

- signal 2
처음 값 : 6초
노란불 : +7초
13 % 7 = 6

서로 만나는 최대값 x
1~x 사이에 13 % 5 == 3 and 13 % 7 == 6 인 값을 찾는다

노란불이 지속된다면?
3~5초 시작 + 7초 -> x % 7 == 3 ~ 5
4초 시작 + 7초 -> x % 7 == 4
3초 시작 + 4초 -> x % 4 == 3

- 문제를 풀고 테스트 결과 보고 든 생각
아마 최대값에 대한 제한 -> 즉, 최소공배수를 설정 안해서 각종 테스트 결과에서 문제가 발생

- 확장 이전 문제 풀이 : limit를 주지 않고 x의 최대값을 임의로 2,000,000 으로 설정함
- 확장 이후 문제 풀이 방법
LCM(최소공배수) 구하는 방법은 GCD(최대공약수)를 구해야 함
int gcd(int a, int b) {
    while(b != 0) {
        int temp = a % b;
        a = b;
        b = temp;
    }
    return a;
}
위 방식을 유클리드 호제법이라고 함
핵심은 두 수 a, b의 최대 공약수는 `b`와 `a를 b로 나눈 나머지`의 최대공약수와 같다
a는 처음 값에서 나누는 값으로 전환됨
b는 나누는 값에서 나머지로 전환됨
그렇게 가다가 나머지가 0이 되면 최대공약수가 정해짐

int lcm(int a, int b) {
    return a * b / gcd(a, b);
}
최소공배수는 서로수를 제외한 공통되는 수가 2번 들어가게 됨
따라서 공통되는 가장 큰 값 = 최대공약수를 한 번 나누어 최소공배수를 구함
*/